### ISSUE TITLE
Implemente front-end extension

### ISSUE THREAD

**Issue by Wh1isper** [Wh1isper/jupyter_kernel_executor#8]
It is better to implement a front-end that replaces (or adds a button) to call the server-side plugin

Currently jupyter can already start the kernel and connect to it automatically, so it is only necessary to maintain the state related to the execution of the backend code

Desired front-end implementation.
- The button triggers a POST interface to initiate an execution request to a cell of the ipynb file
- Poll the GET interface to get the ipynb file and the corresponding cell that the kernel is currently executing, and set it to *.

I lack front-end experience, is there anyone who can assist with this feature?

I have implemented a small demo myself, adding a button to perform a POST operation: https://github.com/Wh1isper/jupyter_kernel_executor/pull/7

**Comment by Wh1isper** [comment]
Due to a bug in Jupyterlab 4.x：https://github.com/jupyterlab/jupyterlab/issues/13606

It is possible for files to be accidentally downgraded, and perhaps we need some temporary measures to protect

**Comment by matthewwiese** [comment]
Would love to lend a hand on this! Should be able to dedicate some time to this next week.

Regarding your second bullet point, do you mean to poll the GET endpoint `<kernel_id>/execute` for active cells and update their output so that the user doesn't have to refresh the page to see updates?

**Comment by Wh1isper** [comment]
@matthewwiese Thank you!

> Regarding your second bullet point, do you mean to poll the GET endpoint <kernel_id>/execute for active cells and update their output so that the user doesn't have to refresh the page to see updates?

Yes, this is for running cells, because currently I haven't found a good way to express the running status on the backend. `execution_count` is `null` or `int` in the `nbformat`, and JupyterLab uses the `*` to indicate that it is tracking running cells, which will disappear after refreshing. So I want the front-end to tell the user **which cells are running** by polling.

The flow I envision is as follows（depends on JupyterLab4 Collaborative）:

0.  (The font-end of this plug-in) POST `<kernel_id>/execute` to execute a cell
1.  (The back-end of this plug-in) Start executing the code
2.  (The font-end of this plug-in) GET `<kernel_id>/execute` for running cells, running cells's `execution_count` should be masked as `*`, otherwise the original `execution_count` is displayed
3.  (The back-end of this plug-in) Write the code execution results to a file
4.  (Collaborative backend) Monitor the file changes and load the file, change ydoc, broadcast to all front ends
5.  (Collaborative front-end) Sync ydoc, refresh widget

You'll notice that in this process, it relies heavily on JupyterLab4's ability to load file changes from local, but as of now, JupyterLab4's experience with these features is not perfect, so we still have a long way to go.

For JupyterLab3, I don't know if polling for notebook content will cause performance or other problems, and there is no indication that JupyterLab3 will have the above features, so I'm not sure what should be done for JupyterLab3. Perhaps we could design a websocket interface that allows the server to push updates of the cell (as far as I know there is no interface to get the output of just one or more cells), but it looks like that's what the collaborative service is trying to do.

Feel free to give your comments. Perhaps there is a better solution😄

**Comment by matthewwiese** [comment]
Hi @Wh1isper!

Sorry for the delay on this, I have been quite busy.

I've recently [pushed a commit](https://github.com/matthewwiese/jupyter_kernel_executor/commit/4da28be10f1cbf3f39f36685e9ef5b5a68dc2691) to the `issue8` branch on my fork of this repo.

It uses WebSockets on the frontend to communicate with the `jupyter_kernel_executor` extension, utilizing your existing code to perform execution, etc. Additionally, Jupyter's JS APIs are used to update the cell output and execution counts as well.

It's still quite rough (I have never worked on Jupyter's backend nor frontend before) but take a look and let me know - we can put together a PR or two to merge changes which you think would be valuable.

**Comment by Wh1isper** [comment]
Thank you @matthewwiese 

It looks like you want to exchange the results of the kernel run with websocket, which doesn't seem to be different from the original design of jupyter. What I'd like to know is how the front-end can tell that its files are up-to-date in such a case (so that the front-end doesn't overwrite the back-end results when it saves them)

**Comment by matthewwiese** [comment]
> It looks like you want to exchange the results of the kernel run with websocket, which doesn't seem to be different from the original design of jupyter.

That was indeed the objective! The Jupyter users I support would benefit greatly if normal cell executions did not get canceled when the network connection was interrupted. The work you have done with this extension was a great springboard to build from. My goal is to utilize the backend you created as a means to replicate Jupyter's UI/UX for running notebooks cells - but be able to ensure they are still run to completion even if, say, the user closes their browser tab.

I understand that your goal with this repository is slightly different, leading to my offer of selectively merging some of the work I've done (such as the frontend code as per this issue's topic). Obviously, if it's just a handful of lines please just copy them! Hopefully it'll save you some time wading through a bunch of random Jupyter codebases like I did. :smile:

> What I'd like to know is how the front-end can tell that its files are up-to-date in such a case (so that the front-end doesn't overwrite the back-end results when it saves them)

Do you mean such as when the page is refreshed mid-background execution? That is, output has been displayed "live" using the WebSockets accurate to the background execution, but when the page is refreshed the background execution is still going but the output is no longer being updated (because the connection has been severed)?

If so, that is indeed a good question and one I haven't given enough thought to yet. My first instinct is that on notebook initialization (i.e. when the page is loaded) we exchange a handful of WebSocket messages with the backend to check if there are already existing background executions being run. If the kernel is already running cells, we update the frontend's cell outputs with the backend's and resume the live-updated outputs.

**Comment by Wh1isper** [comment]
> If so, that is indeed a good question and one I haven't given enough thought to yet. My first instinct is that on notebook initialization (i.e. when the page is loaded) we exchange a handful of WebSocket messages with the backend to check if there are already existing background executions being run. If the kernel is already running cells, we update the frontend's cell outputs with the backend's and resume the live-updated outputs.

This sounds a lot like JupyterLab's collaboration via `yjs`, which is essentially a synchronization and conflict resolution between local file modifications and front-end file modifications

> Obviously, if it's just a handful of lines please just copy them! Hopefully it'll save you some time wading through a bunch of random Jupyter codebases like I did. 

I will see if I can introduce them when I have time. I'm also following the progress of JupyterLab and Jupyter Server-related projects, so maybe we can help the community support this feature natively. You can also continue to work on your ideas, but don't forget to look at what is already available.

Anyway, thank you for your work! @matthewwiese

**Comment by matthewwiese** [comment]
> This sounds a lot like JupyterLab's collaboration via yjs, which is essentially a synchronization and conflict resolution between local file modifications and front-end file modifications

Yes, I have tried JupyterLab's new collaboration feature and, although exceptional, it does not satisfy my use case. For example, if two users are viewing an open notebook with the below cell:

```python
import time

f = 0
while True:
    print(f)
    f += 1
    time.sleep(2)
    if f >= 25:
        break
```

One user executes the cell, and for both their view of the execution is updated in real time. Great! However, if the user who initiated the execution closes their browser tab, for the other user viewing the live update, execution stops.

This makes sense, of course - the user who started execution has "terminated" it when closing the tab. However, I would like execution to continue. So that the other person viewing the notebook (who didn't initiate execution) will be able to continue watching it. This pattern holds true in a single-user scenario: if I close my JupyterLab notebook browser tab, I want the server to continue running my code like a headless data pipeline. The new collaborative feature unfortunately does not support that.

> I'm also following the progress of JupyterLab and Jupyter Server-related projects, so maybe we can help the community support this feature natively.

Me as well! If you learn where in Jupyter's wide portfolio of projects these ideas belong, please @ me here. I'd love to support some of this functionality natively without need for a third party extension if there is sufficient interest by the community (and from the maintainers) to implement it.

**Comment by Wh1isper** [comment]
Yes, JupyterLab3's collaboration feature does not meet the demand, but JupyterLab4's collaboration (using [Jupyter_server_ydoc](https://github.com/jupyter-server/jupyter_server_ydoc) as yjs backend) can already load file changes from local to front-end using `yjs`(https://github.com/jupyter-server/jupyter_server_ydoc/blob/main/jupyter_server_ydoc/handlers.py#L244, `maybe_load_document`), but currently there is no suitable way to output the code run result to file, and loading changes from local file will also cause the whole notebook to re-render, which affects the experience.

In addition, jupyter_server_ydoc uses a similar but different approach to this project to ensure that files can still be located after being renamed or moved (effectively solving the disconnected reconnection problem in collaboration). When multiplayer collaboration is enabled, the JupyterLab4 front-end will first request the `fileid` of the file then connection to the jupyter_server_ydoc plugin via the `fileid`; the solution used by this extension is to monitor the file in the background, because we does not need to maintain a long connection.

I'm thinking of a more harmonious solution that can better fit into Jupyter's current ecology, and I noticed that Jupyter server allows configuration of kernel connection, so maybe that's a good entry point: https://github.com/jupyter-server/jupyter_server/blob/main/jupyter_server/services/kernels/websocket.py#L50
